`pysam` install fix

bcai2 opened the pull request:
Resolves the R client's issues with installing `pysam`  by enforcing `setuptools` to be version `58.0.0` within the R client. Reinstalls `setuptools` if necessary.

 Prevents the `r-reticulate` virtual env from being recreated if it already exists and a proper version of `setuptools` is detected.

lebovic commented on R/zzz.R:
```diff
@@ -35,3 +55,82 @@
+  }
+
+  if (is.null(python_path)) {
+    packageStartupMessage("No compatible version of Python (>=3.6, libpython libraries) found.")
```
A clearer explanation of the issue here may be helpful to a user.

bcai2 replied:
I provided some more detail in the README since I'm not sure how useful a more detailed message would be in the command-line output and/or error message, but I see your point. I'll add another detailed comment or two.

lebovic replied:
My assumption is that people will not read the README, and will instead copy and paste code to try to get it running as quickly as possible. Under that assumption, messages like these are really valuable.

lebovic commented on R/zzz.R:
```diff
@@ -35,3 +55,82 @@
+  }
+
+  # Check the version of setuptools. Reinstall if needed.
+  version_check <- reticulate::py_run_string("reset_setuptools = False
+try:
+  from distutils.version import LooseVersion
+  import setuptools
+
+  setuptools_version = setuptools.__version__
+  reset_setuptools = (LooseVersion(setuptools_version) >= LooseVersion('58.0.2'))
+except ModuleNotFoundError:
+  pass
+")
```
Is there a cleaner way that we could do this? E.g. a separate Python file?

Code in strings is generally undesired for maintainability.

bcai2 replied:
The only tricky part would be directing R to the path of the python file, but I think that should be doable.

lebovic commented on README.Rmd:
```diff
@@ -72,6 +72,23 @@ You can install the development version from [GitHub](https://github.com/) with:
+
+### Python requirements
+The R client requires a version of **Python 3.6 or greater** with 
+`libpython` libraries, accessible via the `$PATH`. 
```
I doubt most of the folks using R will know with `libpython` is. Maybe "with development libraries (e.g. `libpython`)"?

lebovic replied:
This comment still holds

lebovic commented on README.Rmd:
```diff
@@ -72,6 +72,23 @@ You can install the development version from [GitHub](https://github.com/) with:
+The R client requires a version of **Python 3.6 or greater** with 
+`libpython` libraries, accessible via the `$PATH`. 
+
+Most manually installed versions of Python will have the necessary libraries
+installed with Python itself.
```
Do automatically installed versions of Python on Windows and our test Linux have the installed libraries?

bcai2 replied:
This seems to work with Linux on the jupyter notebooks. I'll test with a new instance before merging this in.

lebovic commented on README.Rmd:
```diff
@@ -72,6 +72,23 @@ You can install the development version from [GitHub](https://github.com/) with:
+Most manually installed versions of Python will have the necessary libraries
+installed with Python itself.
+
+For MacOS: The built-in `python` version may not have the necessary libraries.
```
Specifying version number (or even just "recent version") may be helpful, so that people don't unnecessarily try to install Python via brew.

lebovic commented on R/zzz.R:
```diff
@@ -7,14 +7,34 @@
+    if (!python_is_compatible(reticulate_python_path)) {
+      packageStartupMessage("Env r-reticulate has an incompatible Python version. Searching for compatible Python...")
+      python_path <- find_compatible_python()
+      packageStartupMessage("Compatible Python found. Env r-reticulate must be rebuilt.")
```
Does this print even if a compatible Python package is not found?

bcai2 replied:
`find_compatible_python()` throws an error if a compatible package isn't found.

lebovic reviewed (approved):
Looks good 👍 . A few comments.

bcai2 commented:
@lebovic I've made several changes since the last review. This is worth looking over when you have time; if anything comes up, I'll fix/merge them in the next round of PRs.

lebovic commented on R/install.R:
```diff
@@ -0,0 +1,264 @@
+#' @note The Python package interface is stored in the global variable `toolchest_client`.
+#'
+#' @export
+install_toolchest <- function() {
```
Is it correct to say that the user has to run `install_toolchest`? The rest of the comment is assuming the steps to run for the R client right now are:

```
install.packages("devtools")
devtools::install_github("trytoolchest/toolchest-client-r")
library("toolchest")

toolchest::install_toolchest()

Sys.setenv(TOOLCHEST_KEY = "xxx")
toolchest::test(input_path = "./input.fastq", output_path = "./output.txt")
```

--

This is a clear improvement from it not working, but we're still requiring a call to `toolchest::install_toolchest()` after running what I think most users will assume is an install process. Is there a way we can make this happen automatically?

I can think of two processes that might work, but I lack full context:

1. Running `install_toolchest` on `.onLoad`
2. Running `install_toolchest` when running a tool (e.g. `kraken2`), and checking to see if it has been run yet

lebovic commented on R/install.R:
```diff
@@ -0,0 +1,264 @@
+}
+
+python_versions_in_path <- function() {
+  python_regex <- "python(3|3.\\d)?$"
+  if (Sys.info()[["sysname"]] == "Windows") {
+    python_regex <- "python(3|3.\\d)?.exe$"
+  }
```
Oof

bcai2 replied:
Yes; both are listed as "Linux".

lebovic commented on R/install.R:
```diff
@@ -0,0 +1,264 @@
+        "  brew reinstall python",
+        "",
+        "After installation, make sure that Homebrew is on your PATH env var.",
+        "If problems persist, set RETICULATE_PYTHON manually and re-install:",
```
These sound like a scary couple lines for someone whose furthest foray into coding is R.

I see below that you specify how to set RETICULATE_PYTHON manually, but there's a chance that people don't even know what their PATH is or how to check it. Maybe we could link to a GitHub issue where we include these instructions, and say something along the lines of "If you're still having issues, contact us at Toolchest or check out github.com/issue/xxx"?

bcai2 replied:
Sure, I'll add some instructions there. (I don't think line about the PATH is strictly necessary anymore.)

lebovic commented on R/install.R:
```diff
@@ -0,0 +1,264 @@
+  if (env_python != loaded_python) {
+    sysname <- Sys.info()[["sysname"]]
+    if (sysname == "Darwin" || sysname == "Linux") {
+      restart_msg <- paste(
+        "The required Python version differs from the pre-loaded Python.",
+        "Restart R and re-run the installation with:",
+        "",
+        "  toolchest::install_toolchest()",
+        "",
+        "Let Toolchest know if this problem persists.",
+        sep = "\n"
+      )
```
Did you hit this when installing on the new Mac?

In the same vein, what did the actual install process look like on the new Mac?

bcai2 replied:
Yes. `reticulate` needs to load python before it can create a new environment, and it can't "unload" it reliably without restarting the R session, from what I can find. This error is hit when: 
* an existing version of R is loaded
* the `r-reticulate` virtual env does not exist, so reticulate creates it
Any changes that are made (e.g., reinstalling a compatible version of `setuptools`) are only propagated to the loaded version of R. The client is built/run on the `r-reticulate` env, so the install breaks if setuptols needs to be reinstalled.

bcai2 replied:
Note that this case only happens at most once (during the first ever package load, when configuring the virtual env and `RETICULATE_PYTHON`).

lebovic reviewed (commented):
Generally, looks good.

The big open issue for me is requiring `install_toolchest` to be run manually, see comments above.

bcai2 commented:
@lebovic I'm merging this in now; any future fixes will be for future PRs.

lebovic commented:
> > Resolves the R client's issues with installing pysam by enforcing:
> > 
> > * setuptools version 58.0.0 in the python env
> 
> FYI there is a new pysam release that works with setuptools >= 58.0.2.

Whoops, forgot to respond to this thread. Thanks, @jmarshall! We shall use :)
